src/pythia/src/utils/multicall.rs

```diff
@@ -0,0 +1,319 @@
+                });
+            }
+        }
+
+        Err(Error::InvalidOutputType("invalid tokens".into()))
+    }
+
+    fn into_token(self) -> Token {
+        Token::Tuple(vec![
+            Token::Address(self.target),
+            Token::Bytes(self.call_data),
+            Token::Uint(self.gas_limit),
+        ])
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct MulticallResult {
+    pub success: bool,
+    pub used_gas: U256,
+    pub return_data: Vec<u8>,
+}
+
+impl Tokenizable for MulticallResult {
+    fn from_token(token: Token) -> Result<Self, Error>
+    where
+        Self: Sized,
+    {
+        if let Token::Tuple(tokens) = token {
+            if tokens.len() != 3 {
+                return Err(Error::InvalidOutputType("invalid tokens number".into()));
+            }
+
+            if let (Token::Bool(success), Token::Uint(used_gas), Token::Bytes(return_data)) =
+                (tokens[0].clone(), tokens[1].clone(), tokens[2].clone())
+            {
+                return Ok(Self {
+                    success,
+                    used_gas,
+                    return_data,
+                });
+            }
+        }
+
+        Err(Error::InvalidOutputType("invalid tokens".into()))
+    }
+
+    fn into_token(self) -> Token {
+        Token::Tuple(vec![
+            Token::Bool(self.success),
+            Token::Bytes(self.return_data),
+        ])
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct Transfer {
+    pub target: H160,
+    pub value: U256,
+}
+
+impl Tokenizable for Transfer {
+    fn from_token(token: Token) -> std::result::Result<Self, Error>
+    where
+        Self: Sized,
+    {
+        if let Token::Tuple(tokens) = token {
+            if tokens.len() != 2 {
+                return Err(Error::InvalidOutputType("invalid tokens number".into()));
+            }
+
+            if let (Token::Address(target), Token::Uint(value)) =
+                (tokens[0].clone(), tokens[1].clone())
+            {
+                return Ok(Self { target, value });
+            }
+        }
+
+        Err(Error::InvalidOutputType("invalid tokens".into()))
+    }
+
+    fn into_token(self) -> Token {
+        Token::Tuple(vec![Token::Address(self.target), Token::Uint(self.value)])
+    }
+}
+
+pub async fn multicall<T: Transport>(
+    w3: &Web3<T>,
+    chain_id: &Nat,
+    calls: Vec<Call>,
+    gas_price: U256,
+) -> Result<Vec<MulticallResult>> {
+    log!("[{PUBLISHER}] chain: {}, prepering multicall", chain_id);
+    let mut calls = calls;
+    let mut result: Vec<MulticallResult> = vec![];
+    let chain = Chains::get(chain_id)?;
+    let contract_addr = address::to_h160(MULTICALL_CONTRACT_ADDRESS)?;
+    let contract = Contract::from_json(w3.eth(), contract_addr, MULTICALL_ABI)
+        .context(PythiaError::InvalidContractABI)?;
+    let from = canister::pma().await.context(PythiaError::UnableToGetPMA)?;
+    let key_info = web3::key_info();
+    let gas_price = (gas_price / 10) * 12;
```

Comment:
Strange calculations. I've also seen it mutiple times, so move it to separate function